File: MegaHerdt.Models/Models/Reparation.cs
﻿
using MegaHerdt.Models.Models.Identity;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace MegaHerdt.Models.Models
{
    public class Reparation
    {
        [Key]
        public int Id { get; set; }
        [ForeignKey("ReparationState")]
        public int ReparationStateId { get; set; }
        [ForeignKey("User")]
        public string EmployeeId { get; set; }
        [ForeignKey("User")]
        public string ClientId { get; set; }
        public int BillId { get; set; }
        [Required]
        public int Amount { get; set; } //LABOR COSTS
        [Required]
        public DateTime Date { get; set; }
        public string ClientDescription { get; set; }
        public string EmployeeObservation { get; set; } = "";
        public DateTime ApproximateTime { get; set; }
        public bool Enabled { get; set; } = true;
        public ReparationState ReparationState { get; set; }
        public User Client { get; set; }
        public User Employee { get; set; }
        public List<ReparationClaim> ReparationsClaims { get; set; }
        public List<ReparationArticle> ReparationsArticles { get; set; }
        public Bill Bill { get; set; }
        public Guid NumeroTicket { get; set; } = new Guid();
        public bool Facturada { get; set; } = false;
        public float TotalArticleAmount
        {
            get
            {
                float total = 0;
                foreach(var reparationArticle in ReparationsArticles)
                {
                    total += (reparationArticle.ArticlePriceAtTheMoment * reparationArticle.ArticleQuantity);
                }
                return total;
            }
        }
    }
}

File: MegaHerdt.FrontEnd/MegaHerdt/src/app/models/Reparation/Reparation.ts
import { MethodOfPayment } from "src/app/utils/MethodOfPayment";
import { ReparationArticle } from "../Article/ReparationArticle";
import { Bill } from "../Bill/Bill";
import { ReparationClaimDetail } from "../ReparationClaims/ReparationClaimDetail";
import { ReparationState } from "../ReparationState/ReparationState";
import { UserDetail } from "../User/UserDetail";

export class Reparation {
    id: number;
    reparationState: ReparationState;
    employee: UserDetail;
    client: UserDetail;
    amount: number;
    date: Date;
    reparationsArticles: Array<ReparationArticle>;
    bill: Bill;
    reparationsClaims: Array<ReparationClaimDetail>;
    clientDescription: string;
    employeeObservation: string;
    diagnostic: string;
    totalArticleAmount: number; 
    approximateTime: Date;
    numeroTicket:  string;
    facturada:  boolean;
    //Cantidad de pagos de la reparación
    paymentsQuantity: number | null = null;
    // Metodo de pago (Efectivo, debito, credito)
    methodOfPayment: MethodOfPayment | null = null;
    tipoObjeto: string;

    constructor(id:number = 0, reparationState: ReparationState = new ReparationState(), employee: UserDetail=new UserDetail(), client: UserDetail=new UserDetail(), amount: number=0,
        date: Date= new Date(), reparationsArticles: Array<ReparationArticle>=[], bill: Bill= new Bill(), employeeObservation: string='', diagnostic: string='', numeroTicket:  string='',
        facturada:  boolean=false, tipoObjeto: string = '') {
        this.id = id;
        this.reparationState = reparationState;
        this.employee = employee;
        this.client = client;
        this.amount = amount;
        this.date = date;
        this.reparationsArticles = reparationsArticles;
        this.bill = bill;
        this.reparationsClaims = [];
        this.clientDescription = "";
        this.employeeObservation = employeeObservation;
        this.diagnostic = diagnostic;
        this.totalArticleAmount = 0;
        this.approximateTime = new Date();
        this.numeroTicket = numeroTicket;
        this.facturada = facturada;
        this.tipoObjeto = tipoObjeto;
    }

}